tez-dag/src/main/java/org/apache/tez/dag/app/DAGAppMaster.java

```diff
@@ -621,7 +621,7 @@ public synchronized void serviceInit(final Configuration conf) throws Exception
       }
     }
 
-    Executors.newFixedThreadPool(conf.getInt(TezConfiguration.TEZ_AM_DAG_APPCONTEXT_THREAD_COUNT_LIMIT,
+    rawExecutor = Executors.newFixedThreadPool(conf.getInt(TezConfiguration.TEZ_AM_DAG_APPCONTEXT_THREAD_COUNT_LIMIT,
```

Comment:
The newFixedThreadPool() still returns an executor with an unbounded queue (actually it's bounded but the  max capacity of the LinkedBlockingQueue is Integer.MAX_VALUE so practically it's unbounded). 

It won't spin up more threads than specified (100) but if all threads are working then it'll start queueing up tasks in a queue which will practically never be full. Doesn't this still cause OOM?

What would be the desired behaviour when the queue is full? Rejection? Exception or blocking?

Reply:
Hey Attila, you are right, even though the number of threads is bounded, the number of pending tasks is unlimited.
In the first version of the patch I did try to limit that as well -- maybe we could have a limit there as well 
 e.g., number_of_threads * 2 https://github.com/apache/tez/pull/75/commits/8358151b555c036f33f2e131a33a246d3c006c5b#diff-755c0ec043a1800cd6cbf31823a59c8fR630

While testing I realized that as we maintain an __initializerMap__ with  all the vertex initializers in memory, will can still hit memory issues with larger number of inputs dirs. Any thoughts how we could avoid that here? 
https://github.com/apache/tez/blob/9d2b61b576a2421ec4fb813489d896d2b89fcce9/tez-dag/src/main/java/org/apache/tez/dag/app/dag/RootInputInitializerManager.java#L136

Reply:
Ahh, right, that also seems to be a problem. Maybe doing it in batches would work but that complicates the logic further. 

If that too much change, maybe we should go back to the original problem and try to solve it a different way. There are 2 phases
* Creating initializer instances
* Calling the initialize method on the created instance

The 2nd used to be done in the executor's thread, the 1st was moved there as part of TEZ-4170. The reason was that in some cases (like HiveSplitGenerator) instance creation (which imho not supposed be slow) is slow.
An alternative solution I proposed earlier is to move the expensive part out of HiveSplitGenerator's constructor into it's initializers method (which was already executed by the executor's thread even before TEZ-4170). That way it wouldn't be needed to execute the 1st phase by the executor.